apps/homescreen/test/unit/metadata_test.js

```diff
@@ -0,0 +1,205 @@
+    });
+
+    test('should not create object stores with newer versions', () => {
+      metadata.upgradeSchema(onUpgradeNeededEvent(2));
+      assert.equal(names.length, 0);
+    });
+  });
+
+  suite('indexedDB backed methods', () => {
+    var objectStores = {};
+    var deletedValues = [];
+    var attachDBTransaction = objectStoreName => {
+      objectStores = {};
+      deletedValues = [];
+      metadata.db.transaction = (objectStoreNames, IDBTransactionMode) => {
+        return {
+          objectStore: name => {
+            return {
+              put: value => {
+                if (name !== objectStoreName) {
+                  return;
+                }
+                objectStores[name] = value;
+              },
+              delete: value => {
+                if (name !== objectStoreName) {
+                  return;
+                }
+                deletedValues.push(value);
+              },
+              clear: () => {
+                if (name !== objectStoreName) {
+                  return;
+                }
+              },
+              openCursor: () => {
+                return {
+                  set onsuccess(cb) {
+                    if (objectStoreNames.indexOf(name) === -1) {
+                      return;
+                    }
+                    var value = (name === 'order') ? {
+                      id: 'abc/',
+                      order: 0
+                    } : {
+                      id: 'abc/',
+                      icon: 'abc'
+                    };
+
+                    cb({
+                      target: {
+                        result: {
+                          value: value,
+                          continue: () => {
+                          }
+                        }
+                      }
+                    });
+                  }
+                };
+              }
+            };
+          },
+          set oncomplete(cb) {
+            cb();
+          },
+          onerror: () => {
+          }
+        };
+      };
+    };
+
+    setup(() => {
+      metadata.db = {};
+    });
+
+    suite('HomeMetadata#set()', () => {
+      var setMetadata = (id, order, icon) => {
+        metadata.set(
+          [{
+            id: id,
+            order: order,
+            icon: icon
+          }]);
+      };
+
+      test('should ignore apps without id', () => {
+        attachDBTransaction('order');
+        var order = 0, icon = 'abc';
+        setMetadata(undefined, order, icon);
+        assert.equal(Object.keys(objectStores).length, 0);
+      });
+
+      test('should store apps with order', () => {
+        attachDBTransaction('order');
+        var id = 'abc/', order = 0;
+        setMetadata(id, order, undefined);
+        assert.equal(Object.keys(objectStores).length, 1);
+        assert.equal(objectStores.order.id, id);
+        assert.equal(objectStores.order.order, order);
+      });
+
+      test('should store apps with icon', () => {
```

Comment:
This and the test above should probably also check (or there should be a separate test for) that the order and icon remain unchanged when they aren't set, respectively.